File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/3fc93652c831882019e3b86122b70f61.go
package main

import (
	"fmt"
	"math"
	"sort"
)

var N, K int
var A [][]int64
var A_sum []int64
var best int64

func addItem(knapsack []int64, size int, sum int64) {
	for k := K - size; k >= 0; k-- {
		knapsack[k+size] = int64(math.Max(float64(knapsack[k+size]), float64(knapsack[k]+sum)))
	}
}

func recurse(start, end int, knapsack []int64) {
	if start >= end {
		return
	}

	if end-start == 1 {
		sum := int64(0)
		best = int64(math.Max(float64(best), float64(knapsack[K])))

		for prefix := 1; prefix <= len(A[start]); prefix++ {
			sum += A[start][prefix-1]
			best = int64(math.Max(float64(best), float64(sum+knapsack[K-prefix])))
		}

		return
	}

	mid := (start + end) / 2
	state := make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := start; i < mid; i++ {
		addItem(state, len(A[i]), A_sum[i])
	}

	recurse(mid, end, state)
	state = make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := mid; i < end; i++ {
		addItem(state, len(A[i]), A_sum[i])
	}

	recurse(start, mid, state)
}

type DSum struct{}

func (d *DSum) solve() {
	var t int
	fmt.Scan(&N, &K)
	A = make([][]int64, N)
	A_sum = make([]int64, N)

	for i := 0; i < N; i++ {
		fmt.Scan(&t)
		A[i] = make([]int64, t)

		for j := 0; j < t; j++ {
			fmt.Scan(&A[i][j])
		}

		if t > K {
			A[i] = A[i][:K]
		}

		A_sum[i] = 0
		for _, a := range A[i] {
			A_sum[i] += a
		}
	}

	best = 0
	knapsack := make([]int64, K+1)
	recurse(0, N, knapsack)
	fmt.Println(best)
}

func main() {
	var solver DSum
	solver.solve()
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/3fc93652c831882019e3b86122b70f61.java
import java.util.*;
import java.io.*;

public class 3fc93652c831882019e3b86122b70f61xcodeeval_processed_4000.json {
    static int N, K;
    static List<List<Long>> A;
    static long[] A_sum;
    static long best;

    static void addItem(long[] knapsack, int size, long sum) {
        for (int k = K - size; k >= 0; k--) {
            knapsack[k + size] = Math.max(knapsack[k + size], knapsack[k] + sum);
        }
    }

    static void recurse(int start, int end, long[] knapsack) {
        if (start >= end) return;

        if (end - start == 1) {
            long sum = 0;
            best = Math.max(best, knapsack[K]);

            for (int prefix = 1; prefix <= A.get(start).size(); prefix++) {
                sum += A.get(start).get(prefix - 1);
                best = Math.max(best, sum + knapsack[K - prefix]);
            }

            return;
        }

        int mid = (start + end) / 2;
        long[] state = knapsack.clone();

        for (int i = start; i < mid; i++) {
            addItem(state, A.get(i).size(), A_sum[i]);
        }

        recurse(mid, end, state);
        state = knapsack.clone();

        for (int i = mid; i < end; i++) {
            addItem(state, A.get(i).size(), A_sum[i]);
        }

        recurse(start, mid, state);
    }

    static class Solver {
        BufferedReader in;
        PrintWriter out;

        Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());
            A = new ArrayList<>();
            A_sum = new long[N];

            for (int i = 0; i < N; i++) {
                int t = Integer.parseInt(in.readLine());
                List<Long> tempList = new ArrayList<>();
                st = new StringTokenizer(in.readLine());
                for (int j = 0; j < t; j++) {
                    tempList.add(Long.parseLong(st.nextToken()));
                }

                if (t > K) {
                    tempList = tempList.subList(0, K);
                }

                A.add(tempList);
                A_sum[i] = tempList.stream().mapToLong(Long::longValue).sum();
            }

            best = 0;
            long[] knapsack = new long[K + 1];
            recurse(0, N, knapsack);
            out.println(best);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        Solver solver = new Solver(in, out);
        solver.solve();
        out.close();
    }
}
